### PR TITLE
tail/args: Fix parsing when -F is used together with --retry or --follow

### PR DESCRIPTION
**Pull request by Joining7943**
This pr fixes the argument parsing when `-F` is used together with `--retry` or `--follow`.

`tail -F --retry` should not erase the implicit `--follow` option, `tail -F --follow` should not erase the implicit `--retry` flag etc.

### CODE REVIEW

**Comment by Joining7943** [comment]
@tertsdiepraam This pr is ready :)

**Comment by Joining7943** [comment]
@tertsdiepraam Could you review, please?

File: src/uu/tail/src/args.rs [hunk 1]
```diff
@@ -182,19 +182,33 @@ impl Settings {
...
+            (true, None) => (Some(FollowMode::Name), true),
+            (false, Some(value)) if value == "name" => {
+                (Some(FollowMode::Name), matches.get_flag(options::RETRY))
+            }
+            (false, Some(_)) => (
+                Some(FollowMode::Descriptor),
+                matches.get_flag(options::RETRY),
+            ),
+            (false, None) => (None, matches.get_flag(options::RETRY)),
+        };
```
Comment (sylvestre) [inline src/uu/tail/src/args.rs:207]: please document the various cases

Comment (Joining7943) [reply src/uu/tail/src/args.rs]: done

File: src/uu/tail/Cargo.toml [hunk 2]
```diff
@@ -29,6 +29,9 @@ fundu = { workspace=true }
 windows-sys = { workspace=true, features = ["Win32_System_Threading", "Win32_Foundation"] }
 winapi-util = { workspace=true }
 
+[dev-dependencies]
+rstest = "0.17.0"
```
Comment (tertsdiepraam) [inline src/uu/tail/Cargo.toml:33]: Shouldn't this be a workspace dependency?

File: src/uu/tail/src/args.rs [hunk 3]
```diff
@@ -182,19 +182,53 @@ impl Settings {
...
-                None
-            } else if matches.get_one::<String>(options::FOLLOW)
-                == Some(String::from("name")).as_ref()
+        // We're parsing --follow, -F and --retry in one go under the following conditions:
+        // * -F sets --retry and --follow=name (the -F flag is restricted by clap to be set only
+        // once in contrast to gnu's tail which allows this flag to be specified multiple times)
+        // * plain --follow or short -f is the same like specifying --follow=descriptor and can
+        // overwrite itself (specified multiple times)
+        // * specifying --retry together with -F doesn't change the setting of `retry` to true. The
+        // --retry flag itself is restricted by clap to be specified only once (unlike gnu's tail)
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:191]: Can't we match GNU by allowing retry multiple times?

File: src/uu/tail/src/args.rs [hunk 4]
```diff
@@ -182,19 +182,53 @@ impl Settings {
...
+        let (follow, retry) = match (
+            matches.get_flag(options::FOLLOW_RETRY),
+            matches.get_one::<String>(options::FOLLOW),
+        ) {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:195]: Let's see if we can reduce the branches here. Maybe:
```rust
let arg_follow_retry = matches.get_flag(options::FOLLOW_RETRY);
let retry = arg_follow_retry || matches.get_flag(options::RETRY);
let follow = match (
    arg_follow_retry,
    matches
        .get_one::<String>(options::FOLLOW)
        .map(|s| s.as_str()),
) {
    (true, Some(_))
        if matches.index_of(options::FOLLOW_RETRY) > matches.index_of(options::FOLLOW) =>
    {
        Some(FollowMode::Name)
    }
    (_, Some("name")) | (true, None) => Some(FollowMode::Name),
    (_, Some(_)) => Some(FollowMode::Descriptor),
    (false, None) => None,
};
```
(I've omitted the comments here. Those are of course great to keep)

Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:195]: Even more reduced (but the condition gets more complicated):
```rust
let arg_follow_retry = matches.get_flag(options::FOLLOW_RETRY);
let arg_follow = match matches
    .get_one::<String>(options::FOLLOW)
    .map(|s| s.as_str())
{
    Some("name") => Some(FollowMode::Name),
    Some(_) => Some(FollowMode::Descriptor),
    None => None,
};
let retry = arg_follow_retry || matches.get_flag(options::RETRY);
let follow = if arg_follow_retry
    && (arg_follow.is_none()
        || matches.index_of(options::FOLLOW_RETRY) > matches.index_of(options::FOLLOW))
{
    Some(FollowMode::Name)
} else {
    arg_follow
};
```

File: src/uu/tail/Cargo.toml (hunk 2 above)
Comment (Joining7943) [reply src/uu/tail/Cargo.toml:33]: yeah, that's better. I overread in the docs that [dev-dependencies]` can be workspace dependencies, too.

File: src/uu/tail/src/args.rs (hunk 3 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:191]: Sure. That's just the way it was before this pr and I wasn't sure. I think `-F` should be allowed also multiple times

Comment (Joining7943) [reply src/uu/tail/src/args.rs]: To be fully compliant with gnu's tail, I think we should set `clap_builder::builder::Command::args_override_self` to true globally for all arguments and not only for specific ones. Afaik, this is the way how they parse arguments. I could do this in another pr, maybe together with unit testing the arguments.

**Review by tertsdiepraam — APPROVED** [review]
Nice!

**Pull request merged**